File: smartlist/src/main/java/com/smartlist/module/question/QuestionService.java
package com.smartlist.module.question;

import com.smartlist.enums.AssessmentStatus;
import com.smartlist.model.Assessment;
import com.smartlist.model.AssessmentBlock;
import com.smartlist.model.Question;
import com.smartlist.module.assessmentblock.AssessmentBlockRepository;
import com.smartlist.module.question.dto.request.QuestionReqDTO;
import com.smartlist.module.question.dto.request.ReorderReqDTO;
import com.smartlist.module.question.dto.response.QuestionRespDTO;
import com.smartlist.module.question.interfaces.IQuestionStrategy;
import com.smartlist.module.question.questionstrategy.QuestionStrategyFactory;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Service
public class QuestionService {

    private final AssessmentBlockRepository assessmentBlockRepository;
    private final QuestionRepository questionRepository;
    private final ModelMapper modelMapper;

    public QuestionService(final AssessmentBlockRepository assessmentBlockRepository,
                           final QuestionRepository questionRepository, final ModelMapper modelMapper){
        this.assessmentBlockRepository = assessmentBlockRepository;
        this.questionRepository = questionRepository;
        this.modelMapper = modelMapper;
    }

    public List<QuestionRespDTO> getAllByAssessmentBlockId(final Integer assessmentBlockId, final boolean forAssessment) {
        Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findById(assessmentBlockId);
        AssessmentBlock assessmentBlock = optionalAssessmentBlock.orElseThrow(() -> QuestionResponseCodes.INVALID_ASSESSMENT_BLOCK_ID);

        List<Question> questions = this.questionRepository.findByAssessmentBlockWithAnswers(assessmentBlock);
        List<QuestionRespDTO> questionRespDTOS = new ArrayList<>();

        questions.forEach(question -> {
            QuestionRespDTO questionRespDTO = this.modelMapper.map(question, QuestionRespDTO.class);
            questionRespDTO.getAnswerOptions().forEach(answerOptionRespDTO -> {
                if(forAssessment) answerOptionRespDTO.setCorrect(null);
            });
            questionRespDTOS.add(questionRespDTO);
        });
        return questionRespDTOS;
    }

    public QuestionRespDTO getQuestionById(final Integer questionId, final boolean forAssessment) {
        Optional<Question> optionalQuestion = this.questionRepository.findById(questionId);
        Question question = optionalQuestion.orElseThrow(() -> QuestionResponseCodes.INVALID_QUESTION_ID);

        QuestionRespDTO questionRespDTO = this.modelMapper.map(question, QuestionRespDTO.class);
        questionRespDTO.getAnswerOptions().forEach(answerOptionRespDTO -> {
            if(forAssessment) answerOptionRespDTO.setCorrect(null);
        });

        return questionRespDTO;
    }

    public Integer createQuestion(final Integer assessmentBlockId, final QuestionReqDTO questionReqDTO) {
        Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findById(assessmentBlockId);
        AssessmentBlock assessmentBlock = optionalAssessmentBlock.orElseThrow(() -> QuestionResponseCodes.INVALID_ASSESSMENT_BLOCK_ID);

        this.throwIfAssessmentNotEditable(assessmentBlock.getAssessment());

        IQuestionStrategy questionStrategy  = QuestionStrategyFactory.getQuestionStrategy(questionReqDTO);
        questionStrategy.validate();

        Integer sortOrder = this.questionRepository.getMaxSortOrder(assessmentBlockId);
        sortOrder = sortOrder != null ? sortOrder + 1 : 1;

        Question question = new Question(questionReqDTO);
        question.setAssessmentBlock(assessmentBlock);
        question.setSortOrder(sortOrder);

        this.questionRepository.save(question);

        return question.getId();
    }

    public void updateQuestion(final Integer questionId, final QuestionReqDTO questionReqDTO) {
        Question question = findQuestionById(questionId);
        throwIfAssessmentNotEditable(question.getAssessmentBlock().getAssessment());

        IQuestionStrategy questionStrategy = QuestionStrategyFactory.getQuestionStrategy(questionReqDTO);
        questionStrategy.validate();

        Question questionToUpdate = new Question(questionReqDTO);
        questionToUpdate.setId(questionId);
        questionToUpdate.setSortOrder(question.getSortOrder());
        questionToUpdate.setAssessmentBlock(question.getAssessmentBlock());
        this.questionRepository.save(questionToUpdate);
    }

    public void reorderQuestions(final Integer assessmentBlockId, final ReorderReqDTO reorderReqDTO) {
        Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findById(assessmentBlockId);
        AssessmentBlock assessmentBlock = optionalAssessmentBlock.orElseThrow(() -> QuestionResponseCodes.INVALID_ASSESSMENT_BLOCK_ID);

        throwIfAssessmentNotEditable(assessmentBlock.getAssessment());

        if (reorderReqDTO.getIds().isEmpty()) {
            throw QuestionResponseCodes.INVALID_QUESTION_ID_LIST;
        }

        List<Question> questions = this.questionRepository.findByAssessmentBlock(assessmentBlock);
        if (questions.size() != reorderReqDTO.getIds().size()) {
            throw QuestionResponseCodes.INVALID_QUESTION_ID_LIST;
        }

        questions.forEach(ab -> {
            if (!reorderReqDTO.getIds().contains(ab.getId())) {
                throw QuestionResponseCodes.INVALID_QUESTION_ID_LIST;
            }
        });

        for(int i = 0; i < reorderReqDTO.getIds().size(); i++){
            int index = i;
            Optional<Question> optionalQuestion = questions.stream().filter(ab -> Objects.equals(ab.getId(), reorderReqDTO.getIds().get(index))).findFirst();
            if (optionalQuestion.isPresent()) {
                optionalQuestion.get().setSortOrder(i + 1);
            }
        }

        this.questionRepository.saveAll(questions);
    }

    @Transactional
    public void deleteQuestion(final Integer questionId) {
        Question question = findQuestionById(questionId);

        throwIfAssessmentNotEditable(question.getAssessmentBlock().getAssessment());

        this.questionRepository.delete(question);
        this.questionRepository.updateSortOrder(question.getSortOrder(), question.getAssessmentBlock().getId());
    }

    private void throwIfAssessmentNotEditable(final Assessment assessment) {
        if(assessment.getStatus() != AssessmentStatus.DRAFT){
            throw QuestionResponseCodes.ASSESSMENT_ACTION_DENIED;
        }
    }

    private Question findQuestionById(final Integer questionId) {
        Optional<Question> optionalQuestion = this.questionRepository.findById(questionId);
        return optionalQuestion.orElseThrow (() -> QuestionResponseCodes.INVALID_QUESTION_ID);
    }
}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/question/question.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';

import { AnswerOption } from 'src/entities/answer-option.entity';
import { AssessmentBlock } from 'src/entities/assessment-block.entity';
import { Assessment } from 'src/entities/assessment.entity';
import { Question } from 'src/entities/question.entity';
import { AssessmentStatus } from 'src/enums/assessment.status';
import { QuestionType } from 'src/enums/question.type';
import { AssessmentBlockRepository } from '../assessment-block/assessment-block.repository';
import { AnswerOptionRepository } from './answer-option.repository';
import { QuestionReqDto } from './dto/request/question.req.dto';
import { ReorderReqDto } from './dto/request/reorder.req.dto';
import { QuestionRespDto } from './dto/response/question.resp.dto';
import { IQuestionStrategy } from './question-interface/question-interface';
import { getQuestionStrategy } from './question-strategy/question-strategy-factory';
import { QuestionRepository } from './question.repository';
import { QuestionResponseCodes } from './question.response.codes';

@Injectable()
export class QuestionService {
  constructor(
    @InjectRepository(AssessmentBlockRepository)
    private assessmentBlockRepository: AssessmentBlockRepository,
    @InjectRepository(QuestionRepository)
    private questionRepository: QuestionRepository,
    @InjectRepository(AnswerOptionRepository)
    private answerOptionRepository: AnswerOptionRepository,
  ) {}
  async getAllByAssessmentBlockId(
    assessmentBlockId: number,
    forAssessment: boolean,
  ): Promise<QuestionRespDto[]> {
    const assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findByIdWithAssessment(
        assessmentBlockId,
      );
    this.throwIfAssessmentBlockNotDefined(assessmentBlock);

    const questions: Question[] =
      await this.questionRepository.findAllByAssessmentBlockWithAnswers(
        assessmentBlock,
      );

    const questionRespDtos: QuestionRespDto[] = [];

    questions.forEach((question: Question) => {
      questionRespDtos.push(new QuestionRespDto(question, forAssessment));
    });
    return questionRespDtos;
  }

  async getQuestionById(
    questionId: number,
    forAssessment: boolean,
  ): Promise<QuestionRespDto> {
    const question: Question =
      await this.questionRepository.findByIdWithAnswerOptions(questionId);

    this.throwIfQuestionNotDefined(question);

    const questionRespDto: QuestionRespDto = new QuestionRespDto(
      question,
      forAssessment,
    );
    return questionRespDto;
  }

  async createQuestion(
    assessmentBlockId: number,
    questionReqDto: QuestionReqDto,
  ): Promise<number> {
    const assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findByIdWithAssessment(
        assessmentBlockId,
      );

    this.throwIfAssessmentBlockNotDefined(assessmentBlock);
    this.throwIfAssessmentNotEditable(assessmentBlock.assessment);

    const questionStrategy: IQuestionStrategy =
      getQuestionStrategy(questionReqDto);
    questionStrategy.validate();

    let sortOrder: number = await this.questionRepository.getMaxSortOrder(
      assessmentBlockId,
    );
    sortOrder = sortOrder != null ? sortOrder + 1 : 1;

    let question: Question = this.getEntityFromQuestionReqDto(questionReqDto);
    question.assessmentBlock = assessmentBlock;
    question.sortOrder = sortOrder;
    question = await this.questionRepository.save(question);
    return question.id;
  }

  async updateQuestion(questionId: number, questionReqDto: QuestionReqDto) {
    let question = await this.questionRepository.findByIdWithRelations(
      questionId,
    );
    this.throwIfQuestionNotDefined(question);
    this.throwIfAssessmentNotEditable(question.assessmentBlock.assessment);

    const questionStrategy: IQuestionStrategy =
      getQuestionStrategy(questionReqDto);
    questionStrategy.validate();

    await this.answerOptionRepository.remove(question.answerOptions);

    question = this.getEntityFromQuestionReqDto(questionReqDto);
    question.id = questionId;
    question.assessmentBlock = question.assessmentBlock;
    await this.questionRepository.save(question);
  }

  async reorderQuestions(
    assessmentBlockId: number,
    reorderReqDto: ReorderReqDto,
  ) {
    const assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findByIdWithAssessment(
        assessmentBlockId,
      );
    this.throwIfAssessmentBlockNotDefined(assessmentBlock);
    this.throwIfAssessmentNotEditable(assessmentBlock.assessment);

    if (reorderReqDto.ids.length == 0) {
      throw QuestionResponseCodes.INVALID_QUESTION_ID_LIST;
    }

    const questions: Question[] =
      await this.questionRepository.findAllByAssessmentBlock(assessmentBlock);
    if (questions.length != reorderReqDto.ids.length) {
      throw QuestionResponseCodes.INVALID_QUESTION_ID_LIST;
    }

    questions.forEach((ab) => {
      if (reorderReqDto.ids.indexOf(ab.id) == -1) {
        throw QuestionResponseCodes.INVALID_QUESTION_ID_LIST;
      }
    });

    reorderReqDto.ids.forEach((id, index) => {
      const question: Question = questions.find((ab) => ab.id == id);
      if (question != null) {
        question.sortOrder = index + 1;
      }
    });

    await this.questionRepository.save(questions);
  }

  async deleteQuestion(questionId: number) {
    const question = await this.questionRepository.findByIdWithRelations(
      questionId,
    );

    this.throwIfQuestionNotDefined(question);
    this.throwIfAssessmentNotEditable(question.assessmentBlock.assessment);

    await this.questionRepository.remove(question);
    await this.questionRepository.updateSortOrder(
      question.sortOrder,
      question.assessmentBlock.id,
    );
  }

  private throwIfAssessmentBlockNotDefined(assessmentBlock: AssessmentBlock) {
    if (assessmentBlock == null) {
      throw QuestionResponseCodes.INVALID_ASSESSMENT_BLOCK_ID;
    }
  }

  private throwIfAssessmentNotEditable(assessment: Assessment) {
    if (assessment.status != AssessmentStatus.DRAFT) {
      throw QuestionResponseCodes.ASSESSMENT_ACTION_DENIED;
    }
  }

  private throwIfQuestionNotDefined(question: Question) {
    if (!question) {
      throw QuestionResponseCodes.INVALID_QUESTION_ID;
    }
  }

  private getEntityFromQuestionReqDto(
    questionReqDto: QuestionReqDto,
  ): Question {
    const question = new Question();
    question.type = questionReqDto.type as QuestionType;

    question.options = questionReqDto.options;
    question.text = questionReqDto.text;
    question.score = questionReqDto.score;
    question.answerOptions = [];
    questionReqDto.answer_options.forEach((answerOptionDto) => {
      const answerOption = new AnswerOption();
      answerOption.correct = answerOptionDto.correct;
      answerOption.text = answerOptionDto.text;
      answerOption.question = question;
      answerOption.score = answerOptionDto.score;
      question.answerOptions.push(answerOption);
    });
    return question;
  }
}
